pkg/ibmvpc/subnetsIPBlocks.go

```diff
@@ -0,0 +1,237 @@
+	address := subnetsBlocks[subnetCRN].freeAddressesBlocks[blockIndex].FirstIPAddress()
+	return address, subnetsBlocks.removeAddressFromFree(address, subnetCRN, blockIndex)
+}
+func (subnetsBlocks subnetsIPBlocks) subnetBlocks(subnetCRN string) []*ipblock.IPBlock {
+	return subnetsBlocks[subnetCRN].splitByFiltersBlocks
+}
+
+func (subnetsBlocks subnetsIPBlocks) removeAddressFromFree(address, subnetCRN string, blockIndex int) error {
+	addressBlock, err := ipblock.FromIPAddress(address)
+	if err != nil {
+		return err
+	}
+	subnetsBlocks[subnetCRN].freeAddressesBlocks[blockIndex] =
+		subnetsBlocks[subnetCRN].freeAddressesBlocks[blockIndex].Subtract(addressBlock)
+	return nil
+}
+
+// ///////////////////////////////////////////
+// filtersBlocks create a slice of disjoint blocks, split according to the rules blocks, and their sum is allCidr:
+// 1. collect a slice all the acl blocks
+// 2. add to the slice all the gs blocks
+// 3. add to the slice the CidrAll block
+// 3. create a list of disjoint blocks from this slice
+type filtersBlocks map[string][]*ipblock.IPBlock
+
+func getFiltersBlocks(rc *datamodel.ResourcesContainerModel) (filtersBlocks, error) {
+	blocks := filtersBlocks{}
+	if err := blocks.addACLRuleBlocks(rc); err != nil {
+		return nil, err
+	}
+	if err := blocks.addSGRulesBlocks(rc); err != nil {
+		return nil, err
+	}
+	blocks.disjointBlocks()
+	return blocks, nil
+}
+
+func (blocks filtersBlocks) disjointBlocks() {
+	for vpc := range blocks {
+		blocks[vpc] = ipblock.DisjointIPBlocks(blocks[vpc], []*ipblock.IPBlock{ipblock.GetCidrAll()})
+	}
+}
+
+func (blocks filtersBlocks) addACLRuleBlocks(rc *datamodel.ResourcesContainerModel) error {
+	for _, aclObj := range rc.NetworkACLList {
+		for _, rule := range aclObj.Rules {
+			var src, dst *string
+			switch ruleObj := rule.(type) {
+			case *vpc1.NetworkACLRuleItemNetworkACLRuleProtocolAll:
+				src = ruleObj.Source
+				dst = ruleObj.Destination
+			case *vpc1.NetworkACLRuleItemNetworkACLRuleProtocolTcpudp:
+				src = ruleObj.Source
+				dst = ruleObj.Destination
+			case *vpc1.NetworkACLRuleItemNetworkACLRuleProtocolIcmp:
+				src = ruleObj.Source
+				dst = ruleObj.Destination
+			default:
+				return fmt.Errorf("ACL has unsupported type for rule: %s ", *aclObj.Name)
+			}
+			if err := blocks.addBlocks(*aclObj.VPC.CRN, []*string{src, dst}); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (blocks filtersBlocks) addSGRulesBlocks(rc *datamodel.ResourcesContainerModel) error {
+	for _, sgObj := range rc.SecurityGroupList {
+		for _, rule := range sgObj.Rules {
+			var remote, local *string
+			switch ruleObj := rule.(type) {
+			case *vpc1.SecurityGroupRuleSecurityGroupRuleProtocolAll:
+				remote = ruleObj.Remote.(*vpc1.SecurityGroupRuleRemote).CIDRBlock
+				local = ruleObj.Local.(*vpc1.SecurityGroupRuleLocal).CIDRBlock
+			case *vpc1.SecurityGroupRuleSecurityGroupRuleProtocolTcpudp:
```

Comment:
remote can also be `Address` field or `Name` field (another SG) , not just `CIDRBlock` field

Reply:
if it is a another SG, I guess I should skip it, since it will be handled at the other SG, right?
if it is address, that probably means that this address is already allocated, so I create a block, but add a check when allocating address if it free

Reply:
at this stage, i do not have the sg members.
however members will create a block with one reserved IP, so I think it is OK to omit them.
@adisos thoughts?

Reply:
fixed